ert3/data/_record.py

```diff
@@ -91,3 +105,188 @@ def ensure_consistent_record_names(cls, multi_ensemble_record):
+        super()._set_transmitted()
+        self._uri = str(uri)
+        self._mime = mime
+
+    async def _transmit(self, record: Record, mime: str):
+        storage_uri = self._storage_path / self._concrete_key
+        if mime == "application/json":
+            contents = json.dumps(record.data)
+            async with aiofiles.open(storage_uri, mode="w") as f:
+                await f.write(contents)
+        else:
+            if isinstance(record.data, list) and isinstance(record.data[0], bytes):
+                async with aiofiles.open(storage_uri, mode="wb") as f:  # type: ignore
+                    await f.write(record.data[0])  # type: ignore
+            else:
+                raise TypeError(f"unexpected record data type {type(record.data)}")
+        self.set_transmitted(storage_uri, mime)
+
+    async def transmit_data(
+        self,
+        data: Union[List[float], Mapping[int, float], Mapping[str, float], List[bytes]],
+        mime,
+    ) -> None:
+        if self.is_transmitted():
+            raise RuntimeError("Record already transmitted")
+        record = Record(data=data)
+        return await self._transmit(record, mime)
+
+    async def transmit_file(
+        self,
+        file: Path,
+        mime,
+    ) -> None:
+        if self.is_transmitted():
+            raise RuntimeError("File already transmitted")
+        if mime != "application/json":
+            raise NotImplementedError(
+                f"cannot transmit file unless json, was {self._mime}"
+            )
+        async with aiofiles.open(str(file), mode="r") as f:
+            contents = await f.read()
+            record = Record(data=json.loads(contents))
+        return await self._transmit(record, mime)
+
+    async def load(self) -> Record:
+        if self._state != RecordTransmitterState.transmitted:
+            raise RuntimeError("cannot load untransmitted record")
+        if self._mime != "application/json":
+            raise NotImplementedError(
+                f"cannot load record unless json, was {self._mime}"
+            )
+        async with aiofiles.open(str(self._uri)) as f:
+            contents = await f.read()
+            return Record(data=json.loads(contents))
+
+    async def dump(self, location: Path) -> None:
+        if self._state != RecordTransmitterState.transmitted:
+            raise RuntimeError("cannot dump untransmitted record")
+        await _copy(self._uri, str(location))
+
+
+class InMemoryRecordTransmitter(RecordTransmitter):
+    TYPE: RecordTransmitterType = RecordTransmitterType.in_memory
+    # TODO: these fields should be Record, but that does not work until
+    # https://github.com/cloudpipe/cloudpickle/issues/403 has been released.
+    _data: Optional[Any] = None
+    _index: Optional[Any] = None
+
+    def __init__(self, name: str):
+        super().__init__(type_=self.TYPE)
+        self._name = name
+        self._mime = ""
+
+    def set_transmitted(self, record: Record, mime: str):
```

Comment:
Should this be private?

Reply:
Right now it could be, but I think there will be scenarios in the future where things have already been uploaded and ert3 would need to set the state of the transmitter to `transmitted` based on information from the user config.

Reply:
I see... I would vote for then having the function be made public at the time it is needed 🤷🏻